### PR TITLE
Remove dependence on cmark-gfm Swift fork

### PR DESCRIPTION
**Pull request by stackotter**
I am looking to use Parsley in one of my projects, however, Parsley is on quite an old version of `cmark-gfm` and relies on the [brokenhandsio cmark-gfm fork](https://github.com/brokenhandsio/cmark-gfm.git) keeping up-to-date. To make Parsley easy to keep up-to-date I have created a `cmark` target containing all required `cmark-gfm` source files (along with the required copyright notice). This removes reliance on the intermediate brokenhandsio fork.

I have also created a script that can be run to update the `cmark-gfm` version automatically.

## Usage of `update_cmark.sh`

To update to the latest commit of `cmark_gfm`:

```sh
./update_cmark.sh
```

To update to a specific commit of `cmark_gfm` (where `[commit]` is the commit hash):

```sh
./update_cmark.sh [commit]
```

## Breaking changes

I made no breaking changes to the public API of Parsley when making this change. I did however have to change the implementation of `Parsley.html` slightly to work with the newest version of `cmark-gfm`

## Why keep up-to-date?

It is always a good idea to make it easy to keep up-to-date, however, `cmark-gfm` has a pretty current and relevant example of why. Recently, a [bug was found in `cmark-gfm`](https://github.com/github/cmark-gfm/security/advisories/GHSA-mc3g-88wq-6f4x) which meant that parsing a maliciously crafted table could lead to executing arbitrary code of the attacker's choosing on the machine doing the parsing.

### CODE REVIEW

**Comment by kevinrenskers** [comment]
Hey @stackotter, thanks for the PR! 

I think it would actually be better to move this cmark code to its own repository (including the script to easily keep it up to date), and to depend on that from within Parsley. That way it can be updated independently from Parsley, it could be used by other projects, etc. It feels a bit strange to have all that code be in this repo.

Alternatively, maybe you could create a PR for [`brokenhandsio/cmark-gfm`](https://github.com/brokenhandsio/cmark-gfm) to add that update script, so that they have an easier time to keep their repo up to date?

**Comment by kevinrenskers** [comment]
Part of the reason why I am hesitant to add this code to this repo (or even in a different repo on the loopwerk organization), is that it moves the responsibility to keep it up to date to me.

> relies on the brokenhandsio cmark-gfm fork keeping up-to-date

It's not great that they are not keeping up to date, that is very true. But I'm not so sure I want to be in their shoes and periodically check if it needs to be updated. If you'd like to create a repo with an up to date version of cmark-gfm, and are going to keep it up to date, I'd be happy to switch dependencies in Parsley.

**Comment by stackotter** [comment]
I could export the target as a product so that other packages can use it. But if you don't like the idea of having the cmark code in Parsley, I can see where you're coming from.

I'll setup a `swift-cmark-gfm` repository (I don't like the structure of the brokenhandsio fork and I would rather not rely on them keeping it up to date).

And then I'll setup a GitHub action that checks for a new release of cmark-gfm once a week or so, and if there is one it will automatically run the update script and then create a pr. That way it should just always be up to date. I'll make it also automatically add version tags that match cmark-gfm.

Does that sound like a good solution?

Btw, the reason that I'm not just creating a fork and adding a Package.swift, and then rebasing whenever cmark-gfm makes a new release, is because building cmark-gfm relies on some generated files which aren't committed to the repository. So it's not as simple as just rebasing.

**Comment by kevinrenskers** [comment]
Sounds good to me! By the way some of the tests failed, so the output seems to have changed?

**Comment by stackotter** [comment]
Yeah, just noticed that, it seems like some of the extensions are disabled, Ill make sure to correct that when I create the new repository.

**Comment by stackotter** [comment]
Hi Kevin,

I have now separated `cmark-gfm` and the updater script into [a separate repository](https://github.com/stackotter/swift-cmark-gfm). I have also fixed the issue with extensions being disabled, all tests pass now.

File: update_cmark.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#!/bin/sh
```
Comment (kevinrenskers) [inline update_cmark.sh]: I think this script can now be removed?

File: Sources/Parsley/Parsley.swift [hunk 2]
```diff
@@ -8,19 +8,48 @@ public enum MarkdownError: Error {
...
+      throw MarkdownError.conversionFailed
+    }
+    
+    // Render the ast into an html string
+    guard let htmlCString = cmark_render_html(&ast.pointee, options.rawValue, parser.pointee.syntax_extensions) else {
+      throw MarkdownError.conversionFailed
+    }
+    
+    // Free memory
+    cmark_parser_free(parser)
```
Comment (kevinrenskers) [inline Sources/Parsley/Parsley.swift:39]: It's quite a lot of low level code, I'm wondering if it makes sense to include a Swift-y API wrapper inside of your repo, or better to just keep it in Parsley. What do you think?

Comment (stackotter) [reply Sources/Parsley/Parsley.swift:39]: My thoughts on that were that if I start adding a Swift-y API wrapper inside of my repo then it'll start just becoming Parsley. The old repo you were using adding their own C function to perform those low level actions though. I could add a similar function if you don't like the low level code?

File: update_cmark.sh (hunk 1 above)
Comment (stackotter) [reply update_cmark.sh]: Yep, my bad

File: Sources/Parsley/Parsley.swift (hunk 2 above)
Comment (kevinrenskers) [reply Sources/Parsley/Parsley.swift:39]: >  if I start adding a Swift-y API wrapper inside of my repo then it'll start just becoming Parsley

Yeah fair enough :)

**Comment by kevinrenskers** [comment]
Is there actually a reason why `swift-tools-version` is set to `5.3` in your package? That means that Parsley now also has to update its minimum Swift version. Not a huge deal, but I'm surprised 5.3 is required for your package? The brokenhandsio one worked with 5.1.

**Comment by stackotter** [comment]
Oh sorry about that, it auto generated as 5.6 and then I downgraded it to the oldest swift version I've seen people using. I'll change it to 5.1 for consistency

**Comment by kevinrenskers** [comment]
Can you please ping me when you released a new version? Then I can release as well. Thanks!

**Comment by stackotter** [comment]
Yep, in v1.0.2 I have downgraded the minimum swift version to 5.1 :) just let me know if there's anything else

**Comment by kevinrenskers** [comment]
Parsley 0.8.0 is released, and all checks are green again! :)

**Comment by stackotter** [comment]
Awesome! I look forward to using Parsley in my project that needs it (when I eventually get around to working on that project again).

**Comment by kevinrenskers** [comment]
Hey @stackotter,

I noticed that your repo https://github.com/stackotter/swift-cmark-gfm is not actually keeping up to date either. 

> I'll setup a swift-cmark-gfm repository (I don't like the structure of the brokenhandsio fork and I would rather not rely on them keeping it up to date).
> And then I'll setup a GitHub action that checks for a new release of cmark-gfm once a week or so, and if there is one it will automatically run the update script and then create a pr. That way it should just always be up to date. I'll make it also automatically add version tags that match cmark-gfm.

I don't think you've done that? There have been quite a few releases of cmark-gfm (https://github.com/github/cmark-gfm/releases) and I thought the whole idea to rely on your repo rather than brokenhand's repo was to keep it up to date.

**Comment by stackotter** [comment]
Sorry about that, I'll try to look into creating that action in the next week or so, it must've slipped my mind (my main goal was to fix the vuln, so I must've forgotten about the action once that was achieved)

**Comment by stackotter** [comment]
Turns out 'in the next week or so' was tonight lol, I've created an action that updates the repo, checks that it still builds, and then commits. Unfortunately there's no way to automatically run the action when the upstream repo is updated (without creating and hosting a service that uses webhooks and whatnot), but I can give you access to run the workflow if you want (or I'm happy to update it myself whenever asked).

**Comment by kevinrenskers** [comment]
Can you run it once a month or so? I don't know if that's a thing with GitHub Actions 🤔

**Comment by kevinrenskers** [comment]
Will your workflow also create a new tag when done? I'd rather not have Parsley depend on the main branch :)

**Comment by stackotter** [comment]
> Will your workflow also create a new tag when done? I'd rather not have Parsley depend on the main branch :)

Oh yeah that would be possible, what do you reckon in terms of versioning? It'd be difficult to conform to semantic versioning with an automated action (it doesn't necessarily always pull a tagged commit, just the latest, and it would have breaking changes from time to time).

Good idea, I'll get it to run on the first of each month.

**Comment by kevinrenskers** [comment]
Ah so it doesn't download the latest release of cmark-gfm, but just whatever is on the master branch at that moment? I hope they never have their master branch in a broken state in between releases 🤔

**Comment by stackotter** [comment]
> Ah so it doesn't download the latest release of cmark-gfm, but just whatever is on the master branch at that moment? I hope they never have their master branch in a broken state in between releases 🤔

My script can download specific releases, but it's not really designed for automatically updating to the latest release. I need to spend more time on the script to be able to identify the latest release and stuff.

**Pull request merged**